### ISSUE TITLE
Loop rarely deadlocks during destruction on OS X

### ISSUE THREAD

**Issue by brson** [joyent/libuv#799]
```
#include <stdio.h>
#include <uv.h>

int main() {
  printf("running");
  uv_loop_t *l = uv_loop_new();
  uv_loop_delete(l);
  return 0;
}
```

Run this in a loop like `for i in {1..100000}; do ./a.out; done`. It eventually deadlocks.

Thread 1 backtrace:

```
#0  0x00007fff84085a6a in __semwait_signal ()
#1  0x00007fff840ac896 in pthread_join ()
#2  0x000000010000a5be in uv_thread_join ()
#3  0x000000010000f5e7 in uv__platform_loop_delete ()
#4  0x00000001000047c0 in uv__loop_delete ()
#5  0x0000000100001ab4 in uv_loop_delete ()
#6  0x00000001000009fe in main ()
```

Thread 2 backtrace:

```
#0  0x00007fff8404ad7a in mach_msg_trap ()
#1  0x00007fff8404b3ed in mach_msg ()
#2  0x00007fff8261e932 in __CFRunLoopRun ()
#3  0x00007fff8261ddbf in CFRunLoopRunSpecific ()
#4  0x00007fff8261db46 in CFRunLoopRun ()
#5  0x000000010000f5a3 in uv__cf_loop_runner ()
#6  0x000000010000dc78 in uv__thread_start ()
#7  0x00007fff84083fd6 in _pthread_start ()
#8  0x00007fff84083e89 in thread_start ()
```

This is using libuv master. I've seen this on both 10.6.8 and 10.7.4.

**Comment by bnoordhuis** [comment]
Confirmed.

/cc @indutny

**Comment by indutny** [comment]
Can you please try applying following patch:

``` patch
commit d5fa633ef22bd40c81af85dd2ee3882cce3c91c4
Author: Fedor Indutny <fedor.indutny@gmail.com>
Date:   Fri May 17 20:31:39 2013 +0400

    darwin: assume CFRunLoopStop() isn't thread-safe

    Use signaling mechanism for loop termination, because CFRunLoopStop() is
    most likely not a thread-safe function and invoking it from other thread
    may sometimes result in a "dead-lock".

    fix #799

diff --git a/src/unix/darwin.c b/src/unix/darwin.c
index cfbfed2..5435139 100644
--- a/src/unix/darwin.c
+++ b/src/unix/darwin.c
@@ -84,9 +84,8 @@ void uv__platform_loop_delete(uv_loop_t* loop) {
   uv__cf_loop_signal_t* s;

   assert(loop->cf_loop != NULL);
-  CFRunLoopStop(loop->cf_loop);
+  uv__cf_loop_signal(loop, NULL, NULL);
   uv_thread_join(&loop->cf_thread);
-  loop->cf_loop = NULL;

   uv_sem_destroy(&loop->cf_sem);
   uv_mutex_destroy(&loop->cf_mutex);
@@ -145,7 +144,12 @@ void uv__cf_loop_cb(void* arg) {
     item = ngx_queue_head(&split_head);

     s = ngx_queue_data(item, uv__cf_loop_signal_t, member);
-    s->cb(s->arg);
+
+    /* This was a termination signal */
+    if (s->cb == NULL)
+      CFRunLoopStop(loop->cf_loop);
+    else
+      s->cb(s->arg);

     ngx_queue_remove(item);
     free(s);
```

**Comment by brson** [comment]
@indutny Thanks for the quick response. Your patch fixes the problem.

**Comment by indutny** [comment]
@bnoordhuis lgty?

**Comment by indutny** [comment]
@bnoordhuis landed in d5fa633, thanks.
